go/workflow.go

```diff
@@ -0,0 +1,333 @@
+			logger.Error("Unrecoverable error in handling a signal.", "Error", errSignal)
+			return errSignal
+		}
+	}
+
+	// here because of events threshold, but account still active? Continue-As-New
+	if customer.AccountActive && !workflowCanceled {
+		return workflow.NewContinueAsNewError(ctx, customer)
+	}
+
+	logger.Info("Loyalty workflow completed.", "Customer", customer, "WorkflowCanceled", workflowCanceled)
+	return nil
+}
+
+// CustomerWorkflowID generates a Workflow ID based on the given customer ID.
+func CustomerWorkflowID(customerID string) string {
+	return fmt.Sprintf("customer-%v", customerID)
+}
+
+func validateCustomerInfo(customer *CustomerInfo) {
+	if customer.StatusLevel == nil {
+		customer.StatusLevel = StatusLevelForPoints(customer.LoyaltyPoints)
+	}
+	if len(customer.Guests) == 0 {
+		customer.Guests = make([]string, customer.StatusLevel.GuestsAllowed)
+	}
+}
+
+func addGuestToCustomer(customer *CustomerInfo, guestID string) error {
+	guestExists := false
+	for _, g := range customer.Guests {
+		guestExists = g == guestID
+	}
+	// replace first empty slot, or append if none
+	if !guestExists {
+		added := false
+		for i, g := range customer.Guests {
+			if len(g) == 0 {
+				customer.Guests[i] = guestID
+				added = true
+				break
+			}
+		}
+		if !added {
+			customer.Guests = append(customer.Guests, guestID)
+		}
+	}
+
+	return nil
+}
+
+func signalAddPoints(ctx workflow.Context, c workflow.ReceiveChannel, customer *CustomerInfo) error {
+	logger := workflow.GetLogger(ctx)
+	activities := Activities{}
+
+	var pointsToAdd int
+	c.Receive(ctx, &pointsToAdd)
+
+	logger.Info("Adding points to customer account.", "PointsAdded", pointsToAdd)
+	customer.LoyaltyPoints += pointsToAdd
+
+	currentStatusOrd := customer.StatusLevel.Ordinal
+	customer.StatusLevel = StatusLevelForPoints(customer.LoyaltyPoints)
+	newStatusOrd := customer.StatusLevel.Ordinal
+
+	statusChange := newStatusOrd - currentStatusOrd
+
+	if statusChange > 0 {
+		err := workflow.ExecuteActivity(ctx, activities.SendEmail,
+			fmt.Sprintf(EmailPromoted, customer.StatusLevel.Name)).
+			Get(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("error running SendEmail activity for status promotion: %w", err)
+		}
+	} else if statusChange < 0 {
+		err := workflow.ExecuteActivity(ctx, activities.SendEmail,
+			fmt.Sprintf(EmailDemoted, customer.StatusLevel.Name)).
+			Get(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("error running SendEmail activity for status demotion: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func signalInviteGuest(ctx workflow.Context, c workflow.ReceiveChannel, customer *CustomerInfo) error {
+	logger := workflow.GetLogger(ctx)
+	activities := Activities{}
+
+	var emailToSend string
+	var guestID string
+	c.Receive(ctx, &guestID)
+
+	logger.Info("Checking to see if customer has enough status to allow for a guest invite.", "Customer", customer)
+	nonEmptyGuests := 0
+	for _, v := range customer.Guests {
+		if len(v) > 0 {
+			nonEmptyGuests++
+		}
+	}
```

Comment:
Once you change the initialization of the guests slice, you can just use `len(customer.Guests)` to see how many guests